### ISSUE TITLE
Check valid() in siemens-s7_fmt_plug.c

### ISSUE THREAD

**Issue by loverszhaokai** [magnumripper/JohnTheRipper#1094]
In valid() function from line 89 to 91. The right format of siemens-s7 are as follows:
```
"$siemens-s7$1$599fe00cdb61f76cc6e949162f22c95943468acb$..."
 
"$siemens-s7$0$387c1fe4ce97e0e71f5a93b4a9557a947cd40d6c$..."
```
p points to "1" or "0" field.

```
line 89:  outcome = atoi(p);
line 90:  if (outcome != 1 && outcome != 0)
line 91:       goto bail;
```

I think the author maybe want to check the field to be "1" or "0", but the code does not meet the need.
Such as formats also pass the check:
```
"$siemens-s7$AAAAAAA$387c1fe4ce97e0e71f5a93b4a9557a947cd40d6c$..."

"$siemens-s7$###########$387c1fe4ce97e0e71f5a93b4a9557a947cd40d6c$..."
   ```
So I think there is a bug for check the field to be "1" or "0"

**Comment by jfoug** [comment]
Change 

```
if (outcome != 1 && outcome != 0)
```

to 

```
if (outcome != 1 && (outcome != 0 || *p != '0') )
```

**Comment by jfoug** [comment]
NOTE, there may be other usages of atoi in valid that are buggy like this, where 0 is valid, and atoi returns 0 for garbage inputs.

**Comment by loverszhaokai** [comment]
@jfoug  

Do you think the field should "$siemens-s7$1$..." or "$siemens-s7$0$...", there can not be any other case ?

**Comment by magnumripper** [comment]
You can probably find that out yourself with a quick look at binary() and salt() [the latter might be named get_salt()].

**Comment by loverszhaokai** [comment]
@magnumripper  
I found the get_salt() and get_binary() has **nothing** to do with the field. 
So is this field useless?

**Comment by magnumripper** [comment]
That wouldn't be the first time @kholia introduces unused fields... I guess he's the one to answer then :-)

**Comment by magnumripper** [comment]
To his defence it may be that the field may have a future meaning.

**Comment by loverszhaokai** [comment]
@magnumripper  

Should I **close** this issue ?

**Comment by kholia** [comment]
It is almost always better (TM) to have unused fields, they can be hijacked later on to extend the format.

**Comment by loverszhaokai** [comment]
@kholia  

Get it. Thank you.

**Comment by jfoug** [comment]
for this type issue (a set of single digits being all that is valid), and pretty easy to see what is happening, and probably the most CPU efficient way to do this would be something like

if (! ((*p == '0' || *p == '1') && p[1] == 0) )   goto bail;

In this case, only "0" or "1" is valid.  There is no call to atoi, no division, etc.  All code is inline, and pretty easy to see what it is looking for.

I do think there are other 'single' digit checks where atoi() was used. If any of them check for "0" being valid, then they are as wrong as this one was.   I did a bunch of valid changes late last year @frank-dittrich was fuzzing like crazy.  We got a bunch of them improved, but I am 100% sure there are more that still need work.

**Comment by loverszhaokai** [comment]
@jfoug  

Thank you for your reply! I use AFL to fuzz john too, but I have got only one bug. This field is unused currently. So I closed the issue. Thank you.

**Comment by magnumripper** [comment]
That single digit field is actually filled in by pcap2john.py... and it looks like the only sane value is a 1. Look at line 287 on.

**Comment by loverszhaokai** [comment]
@magnumripper  

```
 320    outcome = 0  # XXX we don't know this currently!
 321    for c, r in result.items():
 322        found_something = True  # overkill ;(
 323        sys.stdout.write("%s:$siemens-s7$%s$%s$%s\n" % (os.path.basename(cfg_pcap_file),
 324                                                        outcome, c, r))
```

```
 401    if pckt_84:
 402        outcome = 1
 403    else:
 404        outcome = 0
 405    sys.stdout.write("%s:$siemens-s7$%s$%s$%s\n" % (os.path.basename(cfg_pcap_file),
 406                                                    outcome, challenge, response))
```

Thus, the outcome must be either "0" or "1" ?

**Comment by magnumripper** [comment]
Yes, definitely 0 or 1. But actually it shouldn't even be 0:
```python
   366      # print "found packets indices: pckt_108=%d, pckt_141=%d, pckt_84=%d, pckt_92=%d" % (pckt_108, pckt_141, pckt_84,
   367      #   pckt_92) if pckt_84:
   368      #    print "auth ok"
   369      # else:
   370      #    print "auth bad. for brute we need right auth result. exit"
   371      #    sys.exit()
```
It's just that the code for rejecting them is commented out for some reason.

**Comment by magnumripper** [comment]
But bottom line is you should probably accept 0 or 1 and reject any other value.

**Comment by frank-dittrich** [comment]
> It is almost always better (TM) to have unused fields, they can be hijacked later on to extend the format.

What good is it to not have a canonical hash representation of a particular hash?

```
(bleeding-jumbo)run $ cat hashes
$siemens-s7$0$599fe00cdb61f76cc6e949162f22c95943468acb$002e45951f62602b2f5d15df217f49da2f5379cb
$siemens-s7$0$387c1fe4ce97e0e71f5a93b4a9557a947cd40d6c$d7789feee651559a09e2f2d92b57306d2835e209
$siemens-s7$1$599fe00cdb61f76cc6e949162f22c95943468acb$002e45951f62602b2f5d15df217f49da2f5379cb
$siemens-s7$1$387c1fe4ce97e0e71f5a93b4a9557a947cd40d6c$d7789feee651559a09e2f2d92b57306d2835e209

(bleeding-jumbo)run $ ./john hashes --wordlist=pw
Loaded 4 password hashes with 2 different salts (Siemens-S7 [HMAC-SHA1 32/64])
Will run 8 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
321              (?)
321              (?)
123              (?)
123              (?)
4g 0:00:00:00 DONE (2015-03-10 15:38) 133.3g/s 66.66p/s 133.3c/s 266.6C/s 123..321
Use the "--show" option to display all of the cracked passwords reliably
Session completed
At Program Exit
MemDbg_Validate level 0 checking Passed
(bleeding-jumbo)run $ ./john hashes --wordlist=pw
Loaded 4 password hashes with 2 different salts (Siemens-S7 [HMAC-SHA1 32/64])
Remaining 2 password hashes with 2 different salts
Will run 8 OpenMP threads
Press 'q' or Ctrl-C to abort, almost any other key for status
321              (?)
123              (?)
2g 0:00:00:00 DONE (2015-03-10 15:38) 100.0g/s 100.0p/s 200.0c/s 200.0C/s 123..321
Use the "--show" option to display all of the cracked passwords reliably
Session completed
At Program Exit
MemDbg_Validate level 0 checking Passed
```

I don't mind a placeholder for possible future extensions of the format.
But that placeholder shouldn't have varying contents unless there is a real difference, i.e., this format really supports different versions of Siemens S7 hashes.

**Comment by magnumripper** [comment]
@loverszhaokai I deferred accepting your current PR so you can amend to that commit in the fix_s7_valid branch.

For commit messages, consider always referring to any open issue (eg. "See #1094" or even "Closes #1094" - the latter will make this issue close once I accept the commit).

**Comment by magnumripper** [comment]
Yes @frank-dittrich you show another problem. This must be fixed in split() iirc. We fixed similar problems in commits like 513bbe9a. Not having that field in the first place would have been a lot easier...

**Comment by loverszhaokai** [comment]
@magnumripper  OK. I will test and check the problem.
